packages/vue/src/pinia.ts

```diff
@@ -0,0 +1,87 @@
+import { addBreadcrumb, getClient, getCurrentScope, getGlobalScope } from '@sentry/core';
+import { addNonEnumerableProperty } from '@sentry/utils';
+
+// Inline PiniaPlugin type
+type PiniaPlugin = (context: {
+  store: {
+    $id: string;
+    $state: unknown;
+    $onAction: (callback: (context: { name: string; after: (callback: () => void) => void }) => void) => void;
+  };
+}) => void;
+
+type SentryPiniaPluginOptions = {
+  attachPiniaState?: boolean;
+  actionTransformer: (action: any) => any;
+  stateTransformer: (state: any) => any;
+};
+
+export const createSentryPiniaPlugin: (options?: SentryPiniaPluginOptions) => PiniaPlugin = (
+  options: SentryPiniaPluginOptions = {
+    attachPiniaState: true,
+    actionTransformer: action => action,
+    stateTransformer: state => state,
+  },
+) => {
+  const plugin: PiniaPlugin = ({ store }) => {
+    options.attachPiniaState &&
+      getGlobalScope().addEventProcessor((event, hint) => {
+        try {
+          // Get current timestamp in hh:mm:ss
+          const timestamp = new Date().toTimeString().split(' ')[0];
+          const filename = `pinia_state_${store.$id}_${timestamp}.json`;
+
+          hint.attachments = [
+            ...(hint.attachments || []),
+            {
+              filename,
+              data: JSON.stringify(store.$state),
+            },
+          ];
+        } catch (_) {
+          // empty
+        }
+
+        return event;
+      });
+
+    store.$onAction(context => {
+      context.after(() => {
+        const transformedAction = options.actionTransformer(context.name);
+
+        if (typeof transformedAction !== 'undefined' && transformedAction !== null) {
+          addBreadcrumb({
+            category: 'action',
+            message: transformedAction,
+            level: 'info',
+          });
+        }
+
+        /* Set latest state to scope */
+        const transformedState = options.stateTransformer(store.$state);
+        const scope = getCurrentScope();
+
+        if (typeof transformedState !== 'undefined' && transformedState !== null) {
+          const client = getClient();
+          const options = client && client.getOptions();
+          const normalizationDepth = (options && options.normalizeDepth) || 3; // default state normalization depth to 3
+
+          const newStateContext = { state: { type: 'pinia', value: transformedState } };
+
+          addNonEnumerableProperty(
+            newStateContext,
+            '__sentry_override_normalization_depth__',
+            3 + // 3 layers for `state.value.transformedState
+              normalizationDepth, // rest for the actual state
+          );
+
+          scope.setContext('state', newStateContext);
+        } else {
+          scope.setContext('state', null);
```

Comment:
Is there a way for us not to nuke the _entire_ state context here but only the pinia state 🤔 I know it could be tricky and is a bit edge-casy but may be worth it.

Reply:
Does this update make sense? https://github.com/getsentry/sentry-javascript/pull/13841/commits/4a78c04b9bc9e942e5fcac43d4689168c9687c09

Reply:
yes! perfect